pkg/netscan/discover.go

```diff
@@ -0,0 +1,268 @@
+	devices := make([]DiscoveredDevice, 0)
+	for probeResults := range resultCh {
+		if len(probeResults) == 0 {
+			continue
+		}
+
+		for _, probeResult := range probeResults {
+			dev, err := proto.ConvertProbeResult(probeResult, params)
+			if err != nil {
+				params.Logger.Warnf("issue converting probe result to discovered device: %s", err.Error())
+				continue
+			}
+			devices = append(devices, dev)
+		}
+	}
+	return devices
+}
+
+func handleConnectionInternal(host string, port string, conn net.Conn, params workerParams) {
+	// on udp, the dial is always successful, so don't print
+	if !strings.HasPrefix(params.NetworkProtocol, NetworkUDP) {
+		params.Logger.Debugf("Connection dialed %s://%s:%s", params.NetworkProtocol, host, port)
+	}
+
+	results, err := params.proto.OnConnectionDialed(host, port, conn, params.Params)
+	if err != nil {
+		params.Logger.Debugf(err.Error())
+	} else if len(results) > 0 {
+		params.resultCh <- results
+	}
+}
+
+// probe attempts to make a connection to a specific ip and port list to determine
+// if there is a service listening at that ip+port.
+func probe(host string, ports []string, params workerParams) {
+	port0 := ports[0]
+	addr := host + ":" + port0
+
+	params.Logger.Tracef("Dial: %s", addr)
+	conn, err := net.DialTimeout(params.NetworkProtocol, addr, params.Timeout)
+	if err != nil {
+		params.Logger.Tracef(err.Error())
+		// EHOSTUNREACH specifies that the host is un-reachable or there is no route to host.
+		// EHOSTDOWN specifies that the network or host is down.
+		// If either of these are the error, do not bother probing the host any longer.
+		if errors.Is(err, syscall.EHOSTUNREACH) || errors.Is(err, syscall.EHOSTDOWN) {
+			// quit probing this host
+			return
+		}
+		// all other errors will cause the scanning to continue to the other ports
+	} else {
+		defer conn.Close()
+		handleConnectionInternal(host, port0, conn, params)
+	}
+
+	// if we are only scanning 1 port, nothing left to do
+	if len(ports) == 1 {
+		return
+	}
+
+	// scan the rest of the ports async (we know the host is at least reachable now)
+	// the reason for this is that after the initial set of probes are completed, there should be
+	// fewer hosts left, so we can speed up the process like this.
+	var wg sync.WaitGroup
+	for _, port := range ports[1:] {
+		p := port
+		addr := host + ":" + p
+		wg.Add(1)
+
+		// wrap this code in a func in order to be able to defer the close method within
+		// the for loop and not cause resource exhaustion.
+		go func(p2 string) {
+			defer wg.Done()
+
+			params.Logger.Tracef("Dial: %s", addr)
+			conn2, err := net.DialTimeout(params.NetworkProtocol, addr, params.Timeout)
+			if err != nil {
+				params.Logger.Tracef(err.Error())
+				return
+			}
+			defer conn2.Close()
+			handleConnectionInternal(host, p2, conn2, params)
+		}(p)
+	}
+
+	// wait for the rest of the probes to complete
+	wg.Wait()
+}
+
+// ipWorker pulls uint32s from the ipCh, convert to IPs, filters then ip
```

Comment:
? not sure what my original wording meant to say
```suggestion
// ipWorker pulls uint32s from the ipCh, convert to IPs, filters them
```